src/groth16/aggregate/prove.rs

```diff
@@ -1,512 +1,316 @@
 use digest::Digest;
 use ff::{Field, PrimeField};
 use groupy::{CurveAffine, CurveProjective};
-use itertools::Itertools;
 use rayon::prelude::*;
 use sha2::Sha256;
 
 use super::{
-    inner_product, poly::DensePolynomial, structured_scalar_power, AggregateProof, GIPAProof,
-    GIPAProofWithSSM, MultiExpInnerProductCProof, PairingInnerProductABProof, SRS,
+    commit,
+    commit::{VKey, WKey},
+    compress, inner_product,
+    poly::DensePolynomial,
+    structured_scalar_power, AggregateProof, GipaProof, KZGOpening, ProverSRS, TippMippProof,
 };
 use crate::bls::Engine;
 use crate::groth16::{multiscalar::*, Proof};
 use crate::SynthesisError;
 
 /// Aggregate `n` zkSnark proofs, where `n` must be a power of two.
+/// It implements the algorithm section 5 of the paper.
 pub fn aggregate_proofs<E: Engine + std::fmt::Debug>(
-    ip_srs: &SRS<E>,
+    srs: &ProverSRS<E>,
     proofs: &[Proof<E>],
 ) -> Result<AggregateProof<E>, SynthesisError> {
-    let (ck_1, ck_2) = ip_srs.get_commitment_keys();
-
-    if ck_1.len() != proofs.len() || ck_2.len() != proofs.len() {
+    if !srs.has_correct_len(proofs.len()) {
         return Err(SynthesisError::MalformedSrs);
     }
 
-    let ck_1 = &ck_1;
-    let ck_2 = &ck_2;
-
-    par! {
-        let (a, com_a) = {
-            let a = proofs.iter().map(|proof| proof.a).collect::<Vec<_>>();
-            let com_a = inner_product::pairing::<E>(&a, ck_1);
-            (a, com_a)
-        },
-        let (b, com_b) = {
-            let b = proofs.iter().map(|proof| proof.b).collect::<Vec<_>>();
-            let com_b = inner_product::pairing::<E>(ck_2, &b);
-            (b, com_b)
-        },
-        let (c, com_c) = {
-            let c = proofs.iter().map(|proof| proof.c).collect::<Vec<_>>();
-            let com_c = inner_product::pairing::<E>(&c, ck_1);
-            (c, com_c)
-        }
-    }
+    // We first commit to A B and C - these commitments are what the verifier
+    // will use later to verify the TIPP and MIPP proofs - even though the TIPP
+    // and MIPP proofs doesn't use them directly.
+    let a = proofs.iter().map(|proof| proof.a).collect::<Vec<_>>();
+    let b = proofs.iter().map(|proof| proof.b).collect::<Vec<_>>();
+    // A and B are committed together in this scheme
+    let com_ab = commit::pair::<E>(&srs.vkey, &srs.wkey, &a, &b);
+    let c = proofs.iter().map(|proof| proof.c).collect::<Vec<_>>();
+    let com_c = commit::single_g1::<E>(&srs.vkey, &c);
```

Comment:
why did you remove the parallelization?

Reply:
For quick experimentation I had to change many stuff and there were some errors due to the macro - I put it back :)